src/Common/src/System/Security/Cryptography/Pkcs12Kdf.cs

```diff
@@ -0,0 +1,280 @@
+                    // H(H(H(... H(D || I))))
+                    hash.AppendData(D);
+                    hash.AppendData(I);
+
+                    for (int j = iterationCount; j > 0; j--)
+                    {
+                        if (!hash.TryGetHashAndReset(hashBuf, out int bytesWritten) || bytesWritten != hashBuf.Length)
+                        {
+                            Debug.Fail($"Hash output wrote {bytesWritten} bytes when {hashBuf.Length} was expected");
+                            throw new CryptographicException();
+                        }
+
+                        if (j != 1)
+                        {
+                            hash.AppendData(hashBuf);
+                        }
+                    }
+
+                    // 7.  Concatenate A_1, A_2, ..., A_c together to form a pseudorandom
+                    // bit string, A.
+                    //
+                    // 8.  Use the first n bits of A as the output of this entire process.
+
+                    if (hashBuf.Length >= destination.Length)
+                    {
+                        hashBuf.Slice(0, destination.Length).CopyTo(destination);
+                        return;
+                    }
+
+                    hashBuf.CopyTo(destination);
+                    destination = destination.Slice(hashBuf.Length);
+
+                    // B.  Concatenate copies of A_i to create a string B of length v
+                    // bits(the final copy of Ai may be truncated to create B).
+                    CircularCopy(hashBuf, bBuf);
+
+                    // C.  Treating I as a concatenation I_0, I_1, ..., I_(k-1) of v-bit
+                    // blocks, where k = ceiling(s / v) + ceiling(p / v), modify I by
+                    // setting I_j = (I_j + B + 1) mod 2 ^ v for each j.
+                    for (int j = (I.Length / vBytes) - 1; j >= 0; j--)
+                    {
+                        Span<byte> I_j = I.Slice(j * vBytes, vBytes);
+                        AddPlusOne(I_j, bBuf);
+                    }
+                }
+            }
+            finally
+            {
+                CryptographicOperations.ZeroMemory(I);
+
+                if (IRented != null)
+                {
+                    ArrayPool<byte>.Shared.Return(IRented);
+                }
+
+                hash.Dispose();
+            }
+        }
+
+        private static void AddPlusOne(Span<byte> into, Span<byte> addend)
+        {
+            Debug.Assert(into.Length == addend.Length);
+
+            int carry = 1;
+
+            for (int i = into.Length - 1; i >= 0; i--)
+            {
+                int tmp = carry + into[i] + addend[i];
+                into[i] = (byte)tmp;
+                carry = tmp >> 8;
+            }
+        }
+
+        private static void CircularCopy(ReadOnlySpan<byte> bytes, Span<byte> destination)
+        {
+            while (destination.Length > 0)
```

Comment:
should we fill destination with zeros or something in case `destination.Length < bytes.Length` at first iteration?